### ISSUE TITLE
Global node package URLSearchParams not defined

### ISSUE THREAD

**Issue by RaphaelManke** [winglang/wing#2088]
### I tried this

```ts
bring cloud;
class TestApi {
  extern "./nodeHelpers.js" static inflight _encodeBody(): str;

  init() { 
  }
  inflight encodedBody(): str {
    return TestApi._encodeBody();
  }

   
}

let testApi = new TestApi();

new cloud.Function(inflight (s: str): str => {
  log("------ test: find ------");
    let response = testApi.encodedBody();
  assert(response == "helloWorld");
}) as "test: handle urlencoded body";
```
nodeHelpers.js
```js
exports._encodeBody = () => {
  const data = new URLSearchParams({
    wing: "fly",
  }).toString();
  return data;
};
```

### I expected this:

To get a Urlencoded version of the object.

### Instead, this happened

```bash
fail ┌ api.wsim » root/env0/test: handle urlencoded body
     │ ------ test: find ------
     │ ReferenceError: URLSearchParams is not defined
     │     at exports._encodeBody (<path>/nodeHelpers.js:15:16)
     │     at TestApi._encodeBody (<path>/target/test/api.wsim/.wing/clients/TestApi.inflight.js:5:130)
     │     at TestApi.encodedBody (<path>/target/test/api.wsim/.wing/clients/TestApi.inflight.js:9:29)
     │     at Handler.handle (<path>/target/test/api.wsim/.wing/test__handle_urlencoded_body_c89b2f3c.js:12:35)
     └     at exports.handler (<path>/target/test/api.wsim/.wing/test__handle_urlencoded_body_c89b2f3c.js:26:5)
```

### Is there a workaround?

```js
const url = require("url");
exports._encodeBody = () => {
  const data = new url.URLSearchParams({
    wing: "fly",
  }).toString();
  return data;
};
```

### Component

Compiler

### Wing Version

0.11.4

### Node.js Version

v18.13.0

### Platform(s)

MacOS

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

**Comment by Chriscbr** [comment]
Adding the module either [here](https://github.com/winglang/wing/blob/91dcaec338370bb43c896d57db6869cef22fe265/apps/wing/src/commands/compile.ts#L191) or to our inflight sandboxing mechanism [here](https://github.com/winglang/wing/blob/91dcaec338370bb43c896d57db6869cef22fe265/libs/wingsdk/src/target-sim/function.inflight.ts#L43) might fix it

**Comment by staycoolcall911** [comment]
This `bring`s a broader topic to discussion - are there additional modules we want to support? @eladb  interested to hear what you think

**Comment by polamoros** [comment]
Same problems here with `URL` :)

**Comment by hasanaburayyan** [comment]
@eladb @MarkMcCulloh @staycoolcall911 a week ago I attempted to add the globals into the vm2 sandbox context and ran into some complications that stressed my understanding of JavaScript and kicked my ass 😅 Here were my findings..

1. Several globals cannot be included into the sandbox without creating our own proxies for them such as `Infinity` `undefined` and 'NaN' these will produce an error like: `TypeError: 'set' on proxy: trap returned falsish for property 'Infinity'` More details: https://github.com/patriksimek/vm2/issues/330 
2. (The bigger issue) since vm2 proxies everything, for security purposes it explicitly denies anything that tries to change the prototype of a host object. IE running test on `examples/tests/valid/api_paths_vars.w` results in `VMError: Operation not allowed on contextified object.` due to the call `.setPrototypeOf()` (somewhere in the extern code) is proxied to throw an exception. See more here: https://github.com/patriksimek/vm2/issues/491

I feel like we are fighting best practices with vm2 here, and what we want to achieve with passing globals into sandbox context is maybe not best suited for vm2? 

I played around with some other vm2 settings but just found myself getting more tangled in a mess. 

The code change
```typescript
const excludedGlobals = [
  'Infinity',
  'NaN',
  'undefined',
  'process',
];

for (const k of Object.getOwnPropertyNames(global)) {
  try {
    if (excludedGlobals.includes(k)) {
      continue;
    }
    ctx[k] = (global as any)[k];
  } catch {
    // ignore unresolvable globals
    // see https://github.com/winglang/wing/pull/1923
  }
}

const vm = new NodeVM({
  console: "redirect", // we hijack `console.xxx` in `cloud/function.ts`
  require: {
    external: true,
    builtin: ["*"], // allow using all node modules
    context: "sandbox", // require inside the sandbox (addresses #1871)
  },
  sandbox: {
    $simulator: this.context,
    ...ctx,
  },
  env: {
    ...process.env,
    ...this.env,
  },
  timeout: this.timeout,
});
```

**Comment by eladb** [comment]
Let's brainstorm about this in the offsite.

**Comment by staycoolcall911** [comment]
Noted. We will.

**Comment by eladb** [comment]
I think we should get rid of `vm2` in favor of the normal Node.js `vm` module. This will allow us to easily map all of the globals into the VM because Node's VM doesn't use proxy classes.

The downside of the Node VM (and the reason I went with `vm2`) is that it doesn't support `require`s without breaking the sandbox.

So, what I am suggesting is to restore bundling but do this only at runtime (in the simulator) and not during compilation. The idea is that when I invoke a `cloud.Function` for the first time, the simulator will use `esbuild` to bundle the inflight code (with all the modules) and cache the result. This will result in a slight "cold start" at runtime, but I don't think this would be a problem.

This will also allow using ESM modules out of the box and will help normalize the runtime environments across cloud/sim.

We also discussed that we want to support `--no-bundle` for cloud compilation (e.g. `wing compile -t tf-aws --no-bundle main.w`) in order for the output to be more human-readable (e.g. for snapshots/playground/etc).

So I think we should unify the bundling code paths between sim and cloud so that by default it will always bundle and if `--no-bundle` is provided, the compiler won't bundle. I think in this case, the compiler should emit something like `<entrypoint>.esbuild.json` file next to the inflight entrypoint that will include the esbuild configuration for the bundler, so it will be possible to bundle later with the exact same configuration.

Then, when invoking a function in the simulator, if there's a `.esbuild.json` file, it will use it to bundle the inflight code.
